File: src/store/viewerStore.ts
// ViewerStore — 단일 진실원 (ARCHITECTURE.md 행동 contract).
// React 18 useSyncExternalStore 호환.
//
// Reset 규칙:
//   boot_id 변경      → 전체 reset
//   WS disconnected   → connection만 reset, history 유지
//   connected + hello → 새 session, stateLog/lastIntent reset
//   disconnect()      → connection만 reset

import { useSyncExternalStore } from "react";
import type {
  EspMessage,
  FsmState,
  IntentMessage,
  MotionFailReason,
  StateMessage,
} from "@/types/protocol";
import type { ConnectionState } from "@/ws/ViewerConnection";
import type { MotionEvent } from "@/controller/MotionController";
import type { Pose } from "@/data/pose-engine";
import { nextPose } from "@/data/pose-engine";

const STATE_LOG_LIMIT = 20;
const PROTOCOL_ERR_LIMIT = 8;
const RECENT_MSG_LIMIT = 10;

export type MotionUiStatus = "started" | "completed" | "failed" | "cancelled" | null;

// "현재 또는 마지막 시작된 motion" — terminal(completed/failed/cancelled) 후에도 유지.
// UI consumer는 *currently running*이 필요하면 motionStatus === 'started'로 가드 필수.
// 다음 motion의 'started' event가 오거나 hello session reset 시 갱신/clear.
// (이 의미는 MotionChip이 "DONE/FAILED" 상태에서도 motion 이름을 보여주기 위함.)
export interface ActiveMotion {
  name: string;
  durationMs: number;
  correlationId: string;
  startedAt: number;
}

export interface ViewerStoreState {
  connectionState: ConnectionState;
  bootId: string | null;
  currentState: FsmState | null;
  lastIntent: IntentMessage | null;
  stateLog: StateMessage[];
  protocolErrors: { count: number; recent: string[] };
  // motion (UI consumption)
  activeMotion: ActiveMotion | null;
  motionStatus: MotionUiStatus;
  motionFailReason: MotionFailReason | null;
  currentPose: Pose;
  // dev panel (recent ESP messages, latest 10)
  recentMessages: EspMessage[];
  // Phase 4: brain dual ↔ mcu fallback informational
  // 마지막 error{code:"brain_unreachable"} 도착 시점 (Date.now()). UI 5s 자동 dismiss용.
  lastBrainUnreachableAt: number | null;
}

const INITIAL_STATE: ViewerStoreState = {
  connectionState: { kind: "idle" },
  bootId: null,
  currentState: null,
  lastIntent: null,
  stateLog: [],
  protocolErrors: { count: 0, recent: [] },
  activeMotion: null,
  motionStatus: null,
  motionFailReason: null,
  currentPose: "idle",
  recentMessages: [],
  lastBrainUnreachableAt: null,
};

type Listener = () => void;

export class ViewerStore {
  private state: ViewerStoreState = INITIAL_STATE;
  private listeners = new Set<Listener>();

  // ───── React glue ─────
  subscribe = (l: Listener) => {
    this.listeners.add(l);
    return () => {
      this.listeners.delete(l);
    };
  };

  getSnapshot = (): ViewerStoreState => this.state;

  // ───── connection ─────
  setConnectionState(next: ConnectionState) {
    this.update((s) => {
      const wasConnected = s.connectionState.kind === "connected";
      const newConnected = next.kind === "connected";
      // disconnected 시 history 유지, connection만 reset
      if (next.kind === "disconnected" || next.kind === "rejected") {
        return { ...s, connectionState: next };
      }
      // 새 연결 시작은 그냥 update (hello가 와야 진짜 reset)
      void wasConnected;
      void newConnected;
      return { ...s, connectionState: next };
    });
  }

  // ───── ESP messages ─────
  applyEspMessage(msg: EspMessage) {
    this.recordRecent(msg);

    switch (msg.type) {
      case "hello":
        this.applyHello(msg.boot_id, msg.payload.current_state);
        break;
      case "state":
        this.applyState(msg);
        break;
      case "intent":
        this.update((s) => ({ ...s, lastIntent: msg }));
        break;
      case "motion_started":
      case "motion_completed":
      case "motion_failed":
        // motion lifecycle은 IntentDispatcher가 별도로 store.applyMotionEvent 호출
        break;
      case "error":
        // 에러는 기본 recentMessages로 표시. 특정 code는 별도 mirror 필드 갱신:
        //  - brain_unreachable → lastBrainUnreachableAt (Phase 4 informational toast용)
        if (msg.payload.code === "brain_unreachable") {
          this.update((s) => ({ ...s, lastBrainUnreachableAt: Date.now() }));
        }
        break;
      case "heartbeat":
        // currentState mirror 갱신
        this.update((s) => ({ ...s, currentState: msg.payload.current_state }));
        break;
    }
  }

  private applyHello(bootId: string, currentState: FsmState) {
    this.update((s) => {
      const bootIdChanged = s.bootId !== null && s.bootId !== bootId;
      if (bootIdChanged) {
        // 전체 reset
        return {
          ...INITIAL_STATE,
          connectionState: s.connectionState,
          bootId,
          currentState,
        };
      }
      // 같은 boot_id 또는 첫 hello → session reset (stateLog, lastIntent, brain fallback toast)
      return {
        ...s,
        bootId,
        currentState,
        lastIntent: null,
        stateLog: [],
        activeMotion: null,
        motionStatus: null,
        motionFailReason: null,
        lastBrainUnreachableAt: null,
      };
    });
  }

  private applyState(msg: StateMessage) {
    this.update((s) => ({
      ...s,
      currentState: msg.payload.to,
      stateLog: [...s.stateLog, msg].slice(-STATE_LOG_LIMIT),
    }));
  }

  applyMotionEvent(ev: MotionEvent) {
    this.update((s) => {
      const cur = s.activeMotion;
      switch (ev.type) {
        case "started":
          return {
            ...s,
            activeMotion: {
              name: ev.intent,
              durationMs: ev.expectedMs,
              correlationId: ev.correlationId,
              startedAt: performance.now(),
            },
            motionStatus: "started",
            motionFailReason: null,
          };
        case "completed": {
          if (!cur || cur.correlationId !== ev.correlationId) return s;
          return {
            ...s,
            motionStatus: "completed",
            currentPose: nextPose(s.currentPose, cur.name),
          };
        }
        case "failed": {
          if (!cur || cur.correlationId !== ev.correlationId) return s;
          return {
            ...s,
            motionStatus: "failed",
            motionFailReason: ev.reason,
            // pose는 유지 (실패 시)
          };
        }
        case "cancelled": {
          if (!cur || cur.correlationId !== ev.correlationId) return s;
          // 이미 terminal status면 무시
          if (s.motionStatus === "failed" || s.motionStatus === "completed") return s;
          return {
            ...s,
            motionStatus: "cancelled",
          };
        }
      }
    });
  }

  // ───── protocol errors ─────
  recordProtocolError(message: string) {
    this.update((s) => ({
      ...s,
      protocolErrors: {
        count: s.protocolErrors.count + 1,
        recent: [...s.protocolErrors.recent, message].slice(-PROTOCOL_ERR_LIMIT),
      },
    }));
  }

  private recordRecent(msg: EspMessage) {
    this.update((s) => ({
      ...s,
      recentMessages: [...s.recentMessages, msg].slice(-RECENT_MSG_LIMIT),
    }));
  }

  // 명시적 reset (테스트 / 시퀀스 재시작용)
  reset() {
    this.state = INITIAL_STATE;
    this.emit();
  }

  // ───── internals ─────
  private update(fn: (s: ViewerStoreState) => ViewerStoreState) {
    const next = fn(this.state);
    if (next === this.state) return;
    this.state = next;
    this.emit();
  }

  private emit() {
    this.listeners.forEach((l) => l());
  }
}

// React hook — 전체 state 또는 selector
export function useViewerStore<T>(store: ViewerStore, selector: (s: ViewerStoreState) => T): T;
export function useViewerStore(store: ViewerStore): ViewerStoreState;
export function useViewerStore<T>(
  store: ViewerStore,
  selector?: (s: ViewerStoreState) => T,
): T | ViewerStoreState {
  const sel = selector ?? ((s: ViewerStoreState) => s as unknown as T);
  return useSyncExternalStore(
    store.subscribe,
    () => sel(store.getSnapshot()),
    () => sel(store.getSnapshot()),
  );
}

File: unity/Assets/Scripts/store/ViewerStore.cs
// ViewerStore — viewer/src/store/viewerStore.ts의 C# 1:1 이식.
// 단일 진실원 (TS와 동일 reset 규칙).
//
// Reset 규칙:
//   boot_id 변경      → 전체 reset
//   WS Disconnected   → connection만 reset, history 유지
//   connected + hello → 새 session, stateLog/lastIntent reset (lastBrainUnreachableAt 포함)
//   Disconnect()      → connection만 reset

#nullable enable

using System;
using System.Collections.Generic;
using Nubjuk.Viewer.Controller;
using Nubjuk.Viewer.Types;
using Nubjuk.Viewer.Ws;

namespace Nubjuk.Viewer.Store
{
    public enum MotionUiStatus
    {
        Started,
        Completed,
        Failed,
        Cancelled,
    }

    // "현재 또는 마지막 시작된 motion" — terminal 후에도 유지 (UI Chip 표시용).
    public sealed record ActiveMotion(
        IntentName Name,
        long DurationMs,
        string CorrelationId,
        long StartedAtMs);

    // pose 분류 (TS pose-engine.ts와 동일)
    public enum Pose
    {
        Idle,
        Sit,
        Hand,
    }

    public sealed record ProtocolErrors(int Count, IReadOnlyList<string> Recent);

    public sealed record ViewerStoreState(
        ConnectionState ConnectionState,
        string? BootId,
        FsmState? CurrentState,
        IntentMessage? LastIntent,
        IReadOnlyList<StateMessage> StateLog,
        ProtocolErrors ProtocolErrors,
        ActiveMotion? ActiveMotion,
        MotionUiStatus? MotionStatus,
        MotionFailReason? MotionFailReason,
        Pose CurrentPose,
        IReadOnlyList<EspMessage> RecentMessages,
        long? LastBrainUnreachableAtMs);

    public sealed class ViewerStore
    {
        private const int StateLogLimit = 20;
        private const int ProtocolErrLimit = 8;
        private const int RecentMsgLimit = 10;

        private static readonly ViewerStoreState InitialState = new(
            ConnectionState: new ConnectionState.Idle(),
            BootId: null,
            CurrentState: null,
            LastIntent: null,
            StateLog: Array.Empty<StateMessage>(),
            ProtocolErrors: new ProtocolErrors(0, Array.Empty<string>()),
            ActiveMotion: null,
            MotionStatus: null,
            MotionFailReason: null,
            CurrentPose: Pose.Idle,
            RecentMessages: Array.Empty<EspMessage>(),
            LastBrainUnreachableAtMs: null);

        private ViewerStoreState _state = InitialState;
        private readonly List<Action> _listeners = new();

        public ViewerStoreState GetSnapshot() => _state;

        public IDisposable Subscribe(Action listener)
        {
            _listeners.Add(listener);
            return new SubscriptionImpl(() => _listeners.Remove(listener));
        }

        // ───── connection ─────
        public void SetConnectionState(ConnectionState next)
        {
            // disconnected/rejected: connection만 갱신, history 유지
            Update(s => s with { ConnectionState = next });
        }

        // ───── ESP messages ─────
        public void ApplyEspMessage(EspMessage msg)
        {
            RecordRecent(msg);

            switch (msg)
            {
                case HelloMessage hello:
                    ApplyHello(hello.BootId, hello.Payload.CurrentState);
                    break;
                case StateMessage state:
                    ApplyState(state);
                    break;
                case IntentMessage intent:
                    Update(s => s with { LastIntent = intent });
                    break;
                case MotionStartedMessage:
                case MotionCompletedMessage:
                case MotionFailedMessage:
                    // motion lifecycle은 IntentDispatcher가 별도로 ApplyMotionEvent 호출
                    break;
                case ErrorMessage err:
                    if (err.Payload.Code == ErrorCode.BrainUnreachable)
                    {
                        Update(s => s with { LastBrainUnreachableAtMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() });
                    }
                    break;
                case HeartbeatMessage hb:
                    Update(s => s with { CurrentState = hb.Payload.CurrentState });
                    break;
            }
        }

        private void ApplyHello(string bootId, FsmState currentState)
        {
            Update(s =>
            {
                bool bootIdChanged = s.BootId != null && s.BootId != bootId;
                if (bootIdChanged)
                {
                    return InitialState with
                    {
                        ConnectionState = s.ConnectionState,
                        BootId = bootId,
                        CurrentState = currentState,
                    };
                }
                // 같은 boot_id 또는 첫 hello → session reset
                return s with
                {
                    BootId = bootId,
                    CurrentState = currentState,
                    LastIntent = null,
                    StateLog = Array.Empty<StateMessage>(),
                    ActiveMotion = null,
                    MotionStatus = null,
                    MotionFailReason = null,
                    LastBrainUnreachableAtMs = null,
                };
            });
        }

        private void ApplyState(StateMessage msg)
        {
            Update(s =>
            {
                var newLog = new List<StateMessage>(s.StateLog) { msg };
                if (newLog.Count > StateLogLimit) newLog.RemoveRange(0, newLog.Count - StateLogLimit);
                return s with
                {
                    CurrentState = msg.Payload.To,
                    StateLog = newLog,
                };
            });
        }

        public void ApplyMotionEvent(MotionEvent ev)
        {
            Update(s =>
            {
                var cur = s.ActiveMotion;
                switch (ev)
                {
                    case MotionEvent.Started started:
                        return s with
                        {
                            ActiveMotion = new ActiveMotion(
                                Name: started.Intent,
                                DurationMs: started.ExpectedMs,
                                CorrelationId: started.CorrelationId,
                                StartedAtMs: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()),
                            MotionStatus = MotionUiStatus.Started,
                            MotionFailReason = null,
                        };

                    case MotionEvent.Completed completed:
                        if (cur == null || cur.CorrelationId != completed.CorrelationId) return s;
                        return s with
                        {
                            MotionStatus = MotionUiStatus.Completed,
                            CurrentPose = NextPose(s.CurrentPose, cur.Name),
                        };

                    case MotionEvent.Failed failed:
                        if (cur == null || cur.CorrelationId != failed.CorrelationId) return s;
                        return s with
                        {
                            MotionStatus = MotionUiStatus.Failed,
                            MotionFailReason = failed.Reason,
                            // pose 유지 (실패 시 advance X)
                        };

                    case MotionEvent.Cancelled cancelled:
                        if (cur == null || cur.CorrelationId != cancelled.CorrelationId) return s;
                        // 이미 terminal status면 무시 (race guard)
                        if (s.MotionStatus == MotionUiStatus.Failed || s.MotionStatus == MotionUiStatus.Completed) return s;
                        return s with { MotionStatus = MotionUiStatus.Cancelled };

                    default:
                        return s;
                }
            });
        }

        // pose-engine.ts와 동일: persistent (sit/hand) → adopt, transient (surprise/roll_*) → keep, reset (idle/stand) → idle
        private static Pose NextPose(Pose current, IntentName completedMotion) => completedMotion switch
        {
            IntentName.Sit => Pose.Sit,
            IntentName.Hand => Pose.Hand,
            IntentName.Idle => Pose.Idle,
            IntentName.Stand => Pose.Idle,
            // transient: roll_left, roll_right, surprise → 현재 pose 유지
            _ => current,
        };

        // ───── protocol errors ─────
        public void RecordProtocolError(string message)
        {
            Update(s =>
            {
                var newRecent = new List<string>(s.ProtocolErrors.Recent) { message };
                if (newRecent.Count > ProtocolErrLimit) newRecent.RemoveRange(0, newRecent.Count - ProtocolErrLimit);
                return s with { ProtocolErrors = new ProtocolErrors(s.ProtocolErrors.Count + 1, newRecent) };
            });
        }

        private void RecordRecent(EspMessage msg)
        {
            Update(s =>
            {
                var newRecent = new List<EspMessage>(s.RecentMessages) { msg };
                if (newRecent.Count > RecentMsgLimit) newRecent.RemoveRange(0, newRecent.Count - RecentMsgLimit);
                return s with { RecentMessages = newRecent };
            });
        }

        public void Reset()
        {
            _state = InitialState;
            Emit();
        }

        // ───── internals ─────
        private void Update(Func<ViewerStoreState, ViewerStoreState> fn)
        {
            var next = fn(_state);
            if (ReferenceEquals(next, _state)) return;
            _state = next;
            Emit();
        }

        private void Emit()
        {
            // listener 변경 중 호출 방지를 위해 snapshot 순회
            var snapshot = _listeners.ToArray();
            foreach (var l in snapshot) l();
        }

        private sealed class SubscriptionImpl : IDisposable
        {
            private readonly Action _dispose;
            private bool _disposed;
            public SubscriptionImpl(Action dispose) { _dispose = dispose; }
            public void Dispose()
            {
                if (_disposed) return;
                _disposed = true;
                _dispose();
            }
        }
    }
}
